### ISSUE TITLE
FP in S3801 when `throw` is in child functions

### ISSUE THREAD

**Issue by gabriel-vivas-sonarsource** [SonarSource/SonarJS#3625]
| Label | Description |
| --- | --- |
| Rule | S3801 "Functions should use 'return' consistently"  |
| Subject | FP when child function uses 'throw' is not detected as an exit condition |
| Community | https://community.sonarsource.com/t/false-positive-of-typescript-s3801-with-function-that-never-returns/79581 |

### Explanation

For this rule users expect `throw` statements to be an acceptable termination, since that interrupts the function.

The basic case where the `throw` statement is directly in the scope of the function is currently working well.

```js
function formatDateOrThrow(value) {
    if (value instanceof Date) {
        return value.toISOString();
    }
    throw new TypeError(`Expected valid Date, but got "${typeof value}"`);
}
```

However, when the `throw` happens in a child function, we are not seeing this as a termination.
We are currently raising an issue but ideally we shouldn't.

```js
function formatDateOrThrow(value) {
    if (value instanceof Date) {
        return value.toISOString();
    }
    throwDateFormatError(value); // FP
}

function throwDateFormatError(value) {
    throw new TypeError(`Expected valid Date, but got "${typeof value}"`);
}
```

This is only an issue in JavaScript, or in TypeScript if the return type is not properly declared.
When using TypeScript and properly declaring the return type, we don't have a FP:

```ts
function formatDateOrThrow(value): string | never {
    if (value instanceof Date) {
        return value.toISOString();
    }
    throwDateFormatError(value); // No prob
}

function throwDateFormatError(value) {
    throw new TypeError(`Expected valid Date, but got "${typeof value}"`);
}
```

### Suggestion

For TypeScript, we could improve the rule description to mention that the issue is raised on missing `never` return types.

For JavaScript, we could attempt to find if there is a child function call that throws.
Perhaps TypeScript inference can help since the child function's codomain includes the exception?

**Comment by j-ulrich** [comment]
> When using TypeScript and properly declaring the return type, we don't have a FP [...]

But this is not how `never` is used normally. Normally, `never` is used only when a function can *never* return (hence the name). If it can return in some cases, then `never` is omitted because any function could throw an exception and hence return "nothing".

So, IMO, appending `| never` to the return type is a workaround for the FP of [S3801](https://sonarsource.github.io/rspec/#/rspec/S3801/javascript) in such cases but it's not the proper way to write the code, IMO.
Also note that appending `| never` is also not necessary if the `throw` statement is directly in the function body.
So I would say function which are declared to return `never` should be handled the same way as `throw` statements which should then make it unnecessary to append `| never`.

Also note the special behavior of `never`: `<any type> | never` narrows to `<any type>`. See for example [this playground ](https://www.typescriptlang.org/play?#code/DYUwLgBAzg9gtiAagQwE4C5plQSwHYDmEAPhHiAG4ioDcAUHQGYCueAxmDjHtPCAGKs2ACgCUmKNnxFS5KqggBvOhAg5Gw2AhSpRSlaoipwzVDy1I09VQF8DYABaoYAdzIg3AUVTPUwgEQWOmpQECBwAA5gAJ4AhP6i9HZ0oJDGYKbkACYAynzBALy8CILsYjRAA) demonstrating this behavior: Both variables used there (`someVar` and `returnedSomeVar`) are both considered to have the type `string`.

**Comment by gabriel-vivas-sonarsource** [comment]
I think you should be right :)

Unfortunately, as far as I know, TypeScript will not infer `never` in function declarations, only in function expressions for [backward compatibility reasons](https://stackoverflow.com/questions/40251524/typescript-never-type-inference).

We have an imperfect [workaround](https://github.com/SonarSource/SonarJS/blob/66c14e5d1a68232940d540a19fb89872d41c206d/eslint-bridge/src/linting/eslint/rules/no-inconsistent-returns.ts#L69) in the analysis code, which makes the first example work.

**Comment by Wohops** [comment]
Migrated to Jira. New ticket number: JS-106

**Comment by j-ulrich** [comment]
> Migrated to Jira. New ticket number: [JS-106](https://sonarsource.atlassian.net/browse/JS-106)

But please note my comment above: the suggestion from above "For TypeScript, we could improve the rule description to mention that the issue is raised on missing never return types." (which is now also copied into the JIRA issue) does not align with the way such cause is typically written in TypeScript.
Explicitly mentioning `| never` additionally in the return type does not make much sense.

So one would **not** write the code like this:
```ts
// Unusual TypeScript
function formatDateOrThrow(value): string | never {
    if (value instanceof Date) {
        return value.toISOString();
    }
    throwDateFormatError(value); // No prob
}

function throwDateFormatError(value) {
    throw new TypeError(`Expected valid Date, but got "${typeof value}"`);
}
```
but like this:
```ts
// Typical TypeScript
function formatDateOrThrow(value): string {
    if (value instanceof Date) {
        return value.toISOString();
    }
    throwDateFormatError(value); // FP
}

function throwDateFormatError(value): never {
    throw new TypeError(`Expected valid Date, but got "${typeof value}"`);
}
```
and there we still get the FP at the moment.

So just changing the description of the rule is merely a workaround.

I would have added this comment to the JIRA ticket but I don't have access...

**Comment by Wohops** [comment]
Thanks, @j-ulrich. I updated the Jira ticket with your comment.
